This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [207]:
# get api key from your .env file
import os

API_KEY = ''

print(API_KEY)

Jt8PhYAM6evmC2X9s81y


Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [208]:
# First, import the relevant modules
import requests
import json

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [209]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned

# receive full url 
url = 'https://www.quandl.com/api/v3/'
url_stock_set = 'datasets/FSE/'
url_api_key = 'AFX_X.json?api_key=' + API_KEY
url_time = 'start_date=2020-10-21&end_date=2020-10-21'
full_url = url + url_stock_set + url_api_key + '&'+ url_time


In [210]:
#request url and receive JSON file

afx_stock = requests.get(full_url)
afx_stock

<Response [200]>

In [211]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json = afx_stock.text
json

'{"dataset":{"id":10095370,"dataset_code":"AFX_X","database_code":"FSE","name":"Carl Zeiss Meditec (AFX_X)","description":"Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.\\u003cbr\\u003e\\u003cbr\\u003eTrading System: Xetra\\u003cbr\\u003e\\u003cbr\\u003eISIN: DE0005313704","refreshed_at":"2020-12-01T14:48:09.907Z","newest_available_date":"2020-12-01","oldest_available_date":"2000-06-07","column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"frequency":"daily","type":"Time Series","premium":false,"limit":null,"transform":null,"column_index":null,"start_date":"2020-10-21","end_date":"2020-10-21","data":[["2020-10-21",121.8,121.9,119.9,120.6,null,2148.0,258027.5,null,null,null]],"collapse":null,"order":null,"database_id":6129}}'

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

## 1 Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).

In [212]:
# get url
url_time_2017 = 'start_date=2017-01-02&end_date=2017-12-29'
url_2017 = url + url_stock_set + url_api_key + '&'+ url_time_2017

# get request and json data
afx_stock_2017 = requests.get(url_2017)

afx_stock_2017

<Response [200]>

## 2 Convert the returned JSON object into a Python dictionary.

In [213]:
data = afx_stock_2017.json()
data

{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-12-01T14:48:09.907Z',
  'newest_available_date': '2020-12-01',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2017-01-02',
  'end_date': '2017-12-29',
  'data': [['2017-12-29',
    51.76,
    51.94,
    51.45,
    51.76,
    None,
    34640.0,
    1792304.0,
    None,
    None,
    None],
   ['2017-12-28',
    51.65,
    51.82,
    51.43,
    51.6,
    None

## 3 Calculate what the highest and lowest opening prices were for the stock in this period.

In [214]:
stock_price = data['dataset']['data']
stock_price

[['2017-12-29',
  51.76,
  51.94,
  51.45,
  51.76,
  None,
  34640.0,
  1792304.0,
  None,
  None,
  None],
 ['2017-12-28',
  51.65,
  51.82,
  51.43,
  51.6,
  None,
  40660.0,
  2099024.0,
  None,
  None,
  None],
 ['2017-12-27',
  51.45,
  51.89,
  50.76,
  51.82,
  None,
  57452.0,
  2957018.0,
  None,
  None,
  None],
 ['2017-12-22',
  51.05,
  51.5,
  50.92,
  51.32,
  None,
  71165.0,
  3641949.0,
  None,
  None,
  None],
 ['2017-12-21',
  51.16,
  51.52,
  50.9,
  51.4,
  None,
  120649.0,
  6179433.0,
  None,
  None,
  None],
 ['2017-12-20',
  51.88,
  52.04,
  51.2,
  51.27,
  None,
  50587.0,
  2610258.0,
  None,
  None,
  None],
 ['2017-12-19',
  52.73,
  52.73,
  51.07,
  51.66,
  None,
  137313.0,
  7102361.0,
  None,
  None,
  None],
 ['2017-12-18',
  52.37,
  52.75,
  51.61,
  52.62,
  None,
  129733.0,
  6770499.0,
  None,
  None,
  None],
 ['2017-12-15',
  52.7,
  52.7,
  51.64,
  52.01,
  None,
  204080.0,
  10596319.0,
  None,
  None,
  None],
 ['2017-12-14',
  53.

In [215]:
stock_price_header = data['dataset']['column_names']
stock_price_header

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

From my observation, stock_price is a list of lists. In each list item, there is a list including the information (in order) included in stock_price_header. Based on the questions of this assignment, we will need data in columns **Open**, **High**, **Low**, and **Traded Volume**.

I will create a dictionary of list that contains the data in the above-mentioned columns. The keys will be dates, and the value for the key would be a list: [value of high, value of low, value of Trade Volum]

In [216]:
# identify the index for "Date", "High", "Low", and "Traded Volumes"
date_index = stock_price_header.index('Date')
open_index = stock_price_header.index('Open')
high_index = stock_price_header.index('High')
low_index = stock_price_header.index('Low')
close_index = stock_price_header.index('Close')
tradedv_index = stock_price_header.index('Traded Volume')

In [217]:
# create a new dict to store data
stock_dic = {}

for info in data['dataset']['data']:
    stock_dic[info[date_index]] = [info[open_index], info[high_index], info[low_index], info[close_index], info[tradedv_index]]

stock_dic

{'2017-12-29': [51.76, 51.94, 51.45, 51.76, 34640.0],
 '2017-12-28': [51.65, 51.82, 51.43, 51.6, 40660.0],
 '2017-12-27': [51.45, 51.89, 50.76, 51.82, 57452.0],
 '2017-12-22': [51.05, 51.5, 50.92, 51.32, 71165.0],
 '2017-12-21': [51.16, 51.52, 50.9, 51.4, 120649.0],
 '2017-12-20': [51.88, 52.04, 51.2, 51.27, 50587.0],
 '2017-12-19': [52.73, 52.73, 51.07, 51.66, 137313.0],
 '2017-12-18': [52.37, 52.75, 51.61, 52.62, 129733.0],
 '2017-12-15': [52.7, 52.7, 51.64, 52.01, 204080.0],
 '2017-12-14': [53.11, 53.54, 52.15, 52.67, 132981.0],
 '2017-12-13': [52.64, 53.35, 52.48, 53.09, 128434.0],
 '2017-12-12': [52.29, 53.1, 51.82, 52.43, 87911.0],
 '2017-12-11': [52.28, 52.45, 51.26, 52.14, 71817.0],
 '2017-12-08': [51.5, 52.83, 51.28, 52.12, 109157.0],
 '2017-12-07': [50.89, 51.47, 50.81, 51.47, 48123.0],
 '2017-12-06': [50.8, 51.11, 50.39, 50.89, 88730.0],
 '2017-12-05': [51.21, 51.38, 50.4, 51.25, 83023.0],
 '2017-12-04': [49.5, 51.23, 49.5, 51.14, 94385.0],
 '2017-12-01': [49.52, 50.49, 49.1

In [218]:
# use a for loop to find the highest and lowest opening price

highest_openning = 0
lowest_opening = stock_dic['2017-12-29'][0]
highest_date = ''
lowest_date = ''

for date, values in stock_dic.items():
    
    if values[0]:
        if values[0] > highest_openning:
            highest_openning = values[0]
            highest_date = date

        if values[0] < lowest_opening:
            lowest_opening = values[0]
            lowest_date = date

    
print('On the date {}, AFX had its highest openning {} in 2017'.format(highest_date, highest_openning))
print('On the date {}, AFX had its lowest openning {} in 2017'.format(lowest_date, lowest_opening))

On the date 2017-12-14, AFX had its highest openning 53.11 in 2017
On the date 2017-01-24, AFX had its lowest openning 34.0 in 2017


As mentioned above, the highest opening for AFX in 2017 is \\$53.11, and the lowest is \\$ 34.0.

## 4 What is the largest change in any one day (based on High and Low price)?

In [219]:
largest_change = max(x[1]-x[2] for key,x in stock_dic.items())
largest_change

2.8100000000000023

The largest change (based on high and low price) is \\$2.81.

## 5 What was the largest change between any two days (based on Closing Price)?

In [220]:
closing_list = [x[3] for key,x in stock_dic.items()]

largest_difference = max([closing_list[i+1]-closing_list[i] for i in range(len(closing_list)-1)])
largest_difference

largest_difference

2.559999999999995

The largest change (based on closing price) is \\$2.56.

## 6 What was the average daily trading volume during this year?

In [221]:
sum_volume = sum([x[-1] for key,x in stock_dic.items()])
no_date = len(stock_dic)
average = sum_volume/no_date
average

89124.33725490196

The average daily trading volume in 2017 is 89,124.34.

## 7 (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [222]:
volume_list = [x[-1] for key,x in stock_dic.items()]
volume_list = sorted(volume_list)

volume_len = len(volume_list)

if volume_len%2 ==0:
    median_len = volume_len/2
else:
    median_len = round(volume_len/2)

    
# for index, i need to mins 1 from the location of the median
median_volume = volume_list[median_len -1]
median_volume

76286.0

The medium trading volume in 2017  was 76,286.0.